### ISSUE TITLE
Make HTTP /status endpoint fancier

### ISSUE THREAD

**Issue by quodlibetor** [MaterializeInc/materialize#2095]
Currently it is pretty bare bones, but does report some nice top-line metrics.

It would be better if it had some trivial JS that, if run from a browser, would show "current" values:

* query rate
* average latency
* maybe a chart of query rate over time

Opening this so that we can close #1202 because what it's supposed to track is complete, now.

**Comment by waynr** [comment]
:wave: howdy! I'm starting to look into picking this issue off to gain basic familiarity with code in this repo and want to make sure I understand what should be done before getting started.

First off, do I understand correctly that the purpose of the proposed trivial javascript would be to poll the `/status` endpoint to periodically refresh the displayed values?

As for the current values themselves, you mention "query rate", "average latency", and "query rate over time".  If I understand correctly, all of these should be derivable from existing metrics:

* query rate - a PromQL query that looks something like `rate([mz_query_count](https://github.com/MaterializeInc/materialize/blob/0318db4ef4994612c52b7a269eb1d5ea01bd14f4/src/pgwire/src/metrics.rs#L33)[5m])` (this is query rate calculated over a 5m interval) should work, right?
  * the result of this query would be a scalar value

* query rate over time - not totally clear, but i am guessing this is the query rate graphed over some interval. in this case we could use PromQL that looks something like `rate(mz_query_count[5m])[30m:1m]` which would be the query rate (calculated over 5m intervals) across a 30 minute span with 1-minute resolution.
  * the result should be a range vector (roughly 30 discrete data points separated by 1-minute intervals of the query rate calculated for the 5 minutes preceding each data point)
  * not clear how to display this in the status template -- just as a comma-separated list of values? seems to imply the need for a graphing JS library...

* average latency - it's not clear exactly what is meant by "latency" here. is this referring to command execution latency/duration via the existing `[mz_command_durations](https://github.com/MaterializeInc/materialize/blob/0318db4ef4994612c52b7a269eb1d5ea01bd14f4/src/pgwire/src/metrics.rs#L27)` histogram? in that case, my proposal would be to again go with PromQL queries (but i'll skip discussion of an example query).
  * the result of this query should be a scalar value
  * as long as average latency is being calculated, it might be a good idea to consider additional quantiles

Does all that sound right? I want to make sure of this because the "query rate over time" kind of implies non-trivial dashboard/plotting to visualize the multiple data points calculated by the example PromQL query I proposed. This seems kind of problematic to me since moving in the direction of plotting multiple points runs the risk of turning `/status` into a feature incomplete dashboard. Who exactly would be looking at this rather than an external dashboard utilizing these same metrics?

One thing I would like to propose if dashboard-like behavior actually is the intent -- consider addressing that in a separate issue to introduce a `/dashboard` endpoint that serves an actual dashboard using some kind of open source plotting JS library and URI sub-paths that serve different metrics for that JS to poll.

Anyway, if I were to take this on here is how I imagine I would tackle it (assuming my notes above about the desired values are roughly correct):

* Part 1: implement each of the above-mentioned derived metrics (maybe except for the range vector one)
  * introduce a dependency on a prometheus http client library capable of submitting PromQL queries and parsing the results into useful data structures
  * update [`StatusTemplate`](https://github.com/MaterializeInc/materialize/blob/c3c1c3ee8fb1ea87bd09793ab05de93bda52982e/src/materialized/src/http/metrics.rs#L22) with fields to represent the new derived metrics
  * update `handle_status` to set the new `StatusTemplate` fields
  * update the [status http template](https://github.com/MaterializeInc/materialize/blob/main/src/materialized/src/http/templates/status.html) to show the new derived metrics
  * add very basic in-line javascript to implement polling approach to refresh metrics

* Part 2 (separate issue and PR): implement `/dashboard` endpoint
  * include query rate over time (as i've interpreted it in this comment)
  * include latency quantiles over time (broken down by command)
  * include bytes sent over time  
  * implementation details to be discussed in above-mentioned separate issue

To be clear, I don't necessarily think that building a custom dashboard like this into a service is the best approach -- in my view, it would be better to defer that kind of behavior to services that specialize in it (like grafana). I'm just over thinking it "out loud".

**Comment by benesch** [comment]
Hi @waynr, sorry for the very delayed response!
> To be clear, I don't necessarily think that building a custom dashboard like this into a service is the best approach -- in my view, it would be better to defer that kind of behavior to services that specialize in it (like grafana). I'm just over thinking it "out loud".

We agree! We're retooling Materialize to be primarily a cloud-hosted service (https://cloud.materialize.com), where this sort of visualization will happen in a web UI built atop Materialize. So we're no longer looking to implement this /status endpoint in the `materialized`/`environmentd` binary itself. Going to close this out.
